clusterloader2/pkg/measurement/common/network/controller.go

```diff
@@ -0,0 +1,438 @@
+	clientPodNum, _ = strconv.Atoi(podNumber[0])
+	serverPodNum, _ = strconv.Atoi(podNumber[1])
+	if clientPodNum == serverPodNum && clientPodNum == 1 {
+		return clientPodNum, serverPodNum, OneToOne
+	}
+	if (clientPodNum > serverPodNum) && serverPodNum == 1 {
+		return clientPodNum, serverPodNum, ManyToOne
+	}
+	if clientPodNum == serverPodNum {
+		return clientPodNum, serverPodNum, ManyToMany
+	}
+	return clientPodNum, serverPodNum, Invalid
+}
+
+func executeTest(ratio string, duration int, protocol string) {
+	_, _, ratioType := deriveClientServerPodNum(ratio)
+
+	switch ratioType {
+	case OneToOne:
+		fallthrough
+	case ManyToMany:
+		execNToMTest(duration, protocol)
+	default:
+		klog.Error("Invalid Pod Ratio")
+	}
+}
+
+//execNToMTest executes testcase for N client, M server pods.
+func execNToMTest(duration int, protocol string) {
+	var uniqPodPair uniquePodPair
+
+	go formUniquePodPair(workerPodList)
+
+	for {
+		uniqPodPair = <-podPairCh
+		klog.V(3).Info("Unique PodPair :", uniqPodPair)
+		if uniqPodPair.IsLastPodPair {
+			break
+		}
+		sendReqToSrv(uniqPodPair, protocol, duration)
+		time.Sleep(50 * time.Millisecond)
+		//Get timestamp for first pair and use the same for all
+		if firstClientPodTime == 0 {
+			firstClientPodTime = getTimeStampForPod()
+		}
+		sendReqToClient(uniqPodPair, protocol, duration, firstClientPodTime)
+	}
+
+}
+
+func formUniquePodPair(originalMap map[string][]workerPodData) {
+	var uniqPodPair uniquePodPair
+	lastPodPair := uniquePodPair{IsLastPodPair: true}
+
+	var i = 0
+
+	for {
+		for key, value := range originalMap {
+			unUsedPod, err := getUnusedPod(&value)
+			originalMap[key] = value
+			if err != nil {
+				delete(originalMap, key)
+				continue
+			}
+			i++
+
+			if i == 1 {
+				uniqPodPair.SrcPodIp = unUsedPod.podIp
+				uniqPodPair.SrcPodName = unUsedPod.podName
+			} else if i == 2 {
+				uniqPodPair.DestPodIp = unUsedPod.podIp
+				uniqPodPair.DestPodName = unUsedPod.podName
+				i = 0
+				uniqPodPairList = append(uniqPodPairList, uniqPodPair)
+				podPairCh <- uniqPodPair
+			}
+		}
+		if len(originalMap) == 0 {
+			podPairCh <- lastPodPair
+			break
+		}
+	}
+}
+
+func getUnusedPod(unusedPodList *[]workerPodData) (workerPodData, error) {
+	var unusedPod workerPodData
+	if len(*unusedPodList) == 0 {
+		return unusedPod, errors.New("unused pod list empty")
+	}
+	numOfPods := len(*unusedPodList)
+	//extract last pod of slice
+	unusedPod = (*unusedPodList)[numOfPods-1]
+	*unusedPodList = (*unusedPodList)[:numOfPods-1]
+	return unusedPod, nil
+}
+
+func sendReqToClient(uniqPodPair uniquePodPair, protocol string, duration int, futureTimestamp int64) {
+	klog.V(3).Infof("sending req to cli-pod: %s, protocol: %s, futureTimeStamp: %v, duration: %v",
+		uniqPodPair.SrcPodName, protocol, futureTimestamp, duration)
+	switch protocol {
+	case Protocol_TCP:
+		startWork(uniqPodPair.SrcPodName, httpPathMap[TCP_Client], duration, futureTimestamp, "", uniqPodPair.DestPodIp)
```

Comment:
switch doesn't seem to be needed here - the only difference between these calls are TCP_Client/UDP_Client/HTTP_Client 

Just make a helper method that will chose the correct client from this map based on protocol

Reply:
Same for server

Reply:
modified the code as per comments.